pennylane/devices/qubit_mixed/initialize_state.py

```diff
@@ -0,0 +1,59 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Functions to prepare a state."""
+
+from collections.abc import Iterable
+
+import pennylane as qml
+import pennylane.numpy as np
+from pennylane import math
+
+
+def create_initial_state(
+    # pylint: disable=unsupported-binary-operation
+    wires: qml.wires.Wires | Iterable,
+    prep_operation: qml.operation.StatePrepBase = None,
+    like: str = None,
+):
+    r"""
+    Returns an initial state, defaulting to :math:`\ket{0}` if no state-prep operator is provided.
+
+    Args:
+        wires (Union[Wires, Iterable]): The wires to be present in the initial state
+        prep_operation (Optional[StatePrepBase]): An operation to prepare the initial state
+        like (Optional[str]): The machine learning interface used to create the initial state.
+            Defaults to None
+
+    Returns:
+        array: The initial density matrix (tensor form) of a circuit
+    """
+    num_wires = len(wires)
+    num_axes = (
+        2 * num_wires
+    )  # we initialize the density matrix as the tensor form to keep compatibility with the rest of the module
+    if not prep_operation:
+        state = np.zeros((2,) * num_axes, dtype=complex)
+        state[(0,) * num_axes] = 1
+        return math.asarray(state, like=like)
+
+    # Here, to avoid extending the previous class defined in `StatePrepBase`, we directly call the method `state_vector`. However, this requires some levels of abstract translation between state vectors and density matrices.
+    # The concise explanation is that either state vectors or density matrices are always originally just higher-rank tensors. The only diff is that state vectors are originally of rank num_wires, while density matrices are of rank 2*num_wires. Therefore, we can always define the density matrices as the same wires, appended by a 'shifted' set of wires by num_wires. This idea is also used in the wire sewing technique in the catalyst package.
+    dm_wires = qml.wires.Wires(wires + [w + num_wires for w in wires])
+    density_matrix = prep_operation.state_vector(wire_order=list(dm_wires))
```

Comment:
I'm not sure if this is correct. According to the documentation of `StatePrep`,  a `StatePrep` is created with a state vector like so:
```python
prep_op = qml.StatePrep([1 / np.sqrt(2), 1 / np.sqrt(2)], wires=[0])
```
Applying the current implementation on this operation creates
```pycon
>>> res = create_initial_state(wires=[0], prep_operation=prep_op)
>>> res.reshape((2, 2))
[[0.70710678+0.j 0.        +0.j]
 [0.70710678+0.j 0.        +0.j]]
```

Reply:
Im not sure. All the usage of any qubit methods here is only for convenience such that no more repeated convoluting implementations are introduced, so there might be misalignment with original rules. I will double check tmrw

Reply:
Why is there an extra dim at the front (in the example above, `res` has shape `(1,2,2)`)? Is this necessary?

I also tried with 2 and 3 qubit states and there is something I don't understand. The doc function specifies that it returns the (initial) density matrix of a circuit.

- Shouldn't the trace of the density matrix be always 1?
- Shouldn't the dimension be `2^n X 2^n`, where `n` is the number of qubits?

Reply:
Why is there an extra leading dimension (`res` has shape `(1,2,2)` in the example above)? Is this needed?

I created similar examples for the 2 and 3-qubit cases and there is something I don't understand. The doc of the function specified that it returns the (initial) density matrix of a circuit. Therefore, shouldn't the trace be always 1 (even though we are dealing with a mixed-state device)? Shouldn't the dimension be `2^n x 2^n` where `n` is the number of qubits?
I get an output with the shape `2 x 2 ... 2`

Reply:
@PietropaoloFrisoni the mixed module internally uses a different representation of the density matrix, you can reshape the representation to `(2^n, 2^n)` like I did in my example to get the density matrix we know.
```

Reply:
> Im not sure. All the usage of any qubit methods here is only for convenience such that no more repeated convoluting implementations are introduced, so there might be misalignment with original rules. I will double check tmrw

We shouldn't redefine how `StatePrep` should be created for a desired state. Bottom line is, if a user creates a `StatePrep` like how they always do with any other device, they should be able to initialize a mixed state with the same `StatePrep` operation correctly.

Reply:
Yeah I think I got what you meant. Indeed initially I was also thinking to have another class e.g. `DMPrep` but I realized that it requires modifying some classes defined in `operations` as well as `ops.qubit`; here to avoid such modification I decided to make such a roundabout but I ignored that when it's not private there's possibility users get confused here. 
RN I think the only solution is I create another wrapper here to translate between a DMPrep and StatePrep.
Please hmu if any better ideas!

Reply:
I would say that creating a `DMPrep` is relatively low priority. The purpose of `default.mixed` is to simulate noise, and typically the initial state should still be a pure state, thus can be initialized using a regular `StatePrep`

Reply:
Even if you actually take the time to create a `DMPrep` that accepts a density matrix as its initial state, and if your `initialize_state` function is expecting such a state prep operation that contains a density matrix rather than a state vector, then it shouldn't say `prep_operation: StatePrepBase`, because the function does not work with every state prep operation that inherits from `StatePrep`.